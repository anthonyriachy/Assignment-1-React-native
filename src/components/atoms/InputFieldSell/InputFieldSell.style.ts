import { StyleSheet } from "react-native";
import { ThemeColors } from "../../../constants/theme";
import { globalStyles } from "../../../constants/globalStyles";

export const createStyles=(colors:ThemeColors,isDescription:boolean)=>StyleSheet.create({
    container:{
        flexDirection:'row',
        alignItems:'center',
        justifyContent:'space-between',
    },
    error:{
        color:colors.error,
        fontSize:12,
    },
    label:{
        color:colors.text,
        fontSize:12,
        fontFamily:globalStyles.semiBold,
    },
    input:{
        borderWidth:1,
        color:colors.text,
        borderColor:colors.border,
        borderRadius:8,
        paddingVertical:10,
        paddingHorizontal:15,
        ...(isDescription && {height:120}),
        textAlignVertical:'top',
    },
    description:{
        borderWidth:1,
        borderColor:colors.border,
        borderRadius:20,
        paddingVertical:10,
        paddingHorizontal:15,
        height:200,
    }
})