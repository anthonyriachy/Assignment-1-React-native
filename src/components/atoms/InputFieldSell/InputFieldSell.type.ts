import { TextInputProps } from "react-native";

export type inputFieldSellProps={
    placeholder: string;
    label:string;
    error?: string;
    description?:boolean;
}& TextInputProps;

