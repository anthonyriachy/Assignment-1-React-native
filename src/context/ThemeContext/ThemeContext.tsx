import React, { createContext, useEffect, useState } from 'react';
import { useColorScheme } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ThemeType, getThemeColors } from '../../constants/theme';
import { ThemeContextType } from './ThemeContext.type';

export const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const systemColorScheme = useColorScheme();
  const [theme, setTheme] = useState<ThemeType>(systemColorScheme || 'light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as ThemeType);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    AsyncStorage.setItem('theme', newTheme);
  };

  const value = {
    theme,
    colors: getThemeColors(theme),
    toggleTheme,
    setTheme: (newTheme: ThemeType) => {
      setTheme(newTheme);
      AsyncStorage.setItem('theme', newTheme);
    },
  };

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
};

