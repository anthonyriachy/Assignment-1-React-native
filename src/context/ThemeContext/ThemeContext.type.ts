import { getThemeColors, ThemeType } from '../../constants/theme';

export type ThemeContextType = {
    theme: ThemeType;
    colors: ReturnType<typeof getThemeColors>;
    toggleTheme: () => void;
    setTheme: (theme: ThemeType) => void;
  }