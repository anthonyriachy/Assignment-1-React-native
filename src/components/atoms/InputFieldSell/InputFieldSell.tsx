import React from 'react';
import { View, TextInput } from "react-native";
import { inputFieldSellProps } from "./InputFieldSell.type";
import { useTheme } from "../../../hooks/UseTheme";
import { createStyles } from "./InputFieldSell.style";
import { CustomText } from "../CustomText/CustomText";
import { ErrorText } from "../ErrorText/ErrorText";
export const InputFieldSell = ({error,label,description=false,...props}:inputFieldSellProps) => {
    const {colors}=useTheme()
    const styles=createStyles(colors,description)
    return (
        <View>
            <CustomText style={styles.label}>{label}</CustomText>
            {error && <ErrorText error={error} />}
            <TextInput
                placeholderTextColor={colors.placeholder}
                {...props}
                style={styles.input}
                multiline={description}
                numberOfLines={description ? 5 : 1}
            />
        </View>

    )
}