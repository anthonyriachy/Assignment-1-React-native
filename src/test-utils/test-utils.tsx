import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from '../context/ThemeContext/ThemeContext';

// Mock theme hook
jest.mock('../hooks/UseTheme', () => ({
  useTheme: () => ({
    colors: {
      background: '#fff',
      text: '#000',
      primary: '#007AFF',
      border: '#ccc',
    },
    theme: 'light',
  }),
}));

// Custom render function that includes providers
const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: ThemeProvider, ...options });

// Re-export everything
export * from '@testing-library/react-native';

// Override render method
export { customRender as render }; 