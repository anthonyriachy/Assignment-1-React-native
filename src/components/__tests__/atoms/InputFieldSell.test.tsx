import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { render } from '../../../../src/test-utils/test-utils';
import { InputFieldSell } from '../../atoms/InputFieldSell/InputFieldSell';
import { ThemeProvider } from '../../../context/ThemeContext/ThemeContext';

jest.mock('../../atoms/CustomText/CustomText', () => {
  const React = require('react');
  const { Text } = require('react-native');
  return {
    CustomText: ({ children, style }: any) => (
      <Text style={style}>{children}</Text>
    ),
  };
});

jest.mock('../../atoms/ErrorText/ErrorText', () => {
  const React = require('react');
  const { Text } = require('react-native');
  return {
    ErrorText: ({ error }: any) => (
      <Text testID="error-text">{error}</Text>
    ),
  };
});

jest.mock('../../../hooks/UseTheme', () => ({
  useTheme: () => ({
    colors: {
      background: '#fff',
      text: '#000',
      primary: '#007AFF',
      border: '#ccc',
    },
  }),
}));

describe('InputFieldSell', () => {
  const mockOnChangeText = jest.fn();

  beforeEach(() => {
    mockOnChangeText.mockClear();
  });

  it('renders correctly with label', () => {
    const { getByText } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter text"
          onChangeText={mockOnChangeText}
        />
      </ThemeProvider>
    );

    expect(getByText('Test Label')).toBeTruthy();
  });

  it('displays error message when error prop is provided', () => {
    const errorMessage = 'This is an error';
    const { getByTestId } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter text"
          error={errorMessage}
          onChangeText={mockOnChangeText}
        />
      </ThemeProvider>
    );

    expect(getByTestId('error-text')).toHaveTextContent(errorMessage);
  });

  it('handles text input correctly', () => {
    const { getByPlaceholderText } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter text"
          onChangeText={mockOnChangeText}
        />
      </ThemeProvider>
    );

    const input = getByPlaceholderText('Enter text');
    fireEvent.changeText(input, 'test input');
    expect(mockOnChangeText).toHaveBeenCalledWith('test input');
  });

  it('renders in description mode when description prop is true', () => {
    const { getByPlaceholderText } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter description"
          description={true}
          onChangeText={mockOnChangeText}
        />
      </ThemeProvider>
    );

    const input = getByPlaceholderText('Enter description');
    expect(input.props.multiline).toBe(true);
    expect(input.props.numberOfLines).toBe(5);
  });

  it('renders correctly with default props', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter text"
          onChangeText={jest.fn()}
        />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with error', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter text"
          onChangeText={jest.fn()}
          error="Error message"
        />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly in description mode', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <InputFieldSell 
          label="Test Label"
          placeholder="Enter description"
          onChangeText={jest.fn()}
          description={true}
        />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
}); 