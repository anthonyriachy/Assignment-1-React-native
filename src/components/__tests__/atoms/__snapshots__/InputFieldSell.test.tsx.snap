// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFieldSell renders correctly in description mode 1`] = `
<View>
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 12,
      }
    }
  >
    Test Label
  </Text>
  <TextInput
    multiline={true}
    numberOfLines={5}
    onChangeText={[MockFunction]}
    placeholder="Enter description"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "color": "#000",
        "height": 120,
        "paddingHorizontal": 15,
        "paddingVertical": 10,
        "textAlignVertical": "top",
      }
    }
  />
</View>
`;

exports[`InputFieldSell renders correctly with default props 1`] = `
<View>
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 12,
      }
    }
  >
    Test Label
  </Text>
  <TextInput
    multiline={false}
    numberOfLines={1}
    onChangeText={[MockFunction]}
    placeholder="Enter text"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "color": "#000",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
        "textAlignVertical": "top",
      }
    }
  />
</View>
`;

exports[`InputFieldSell renders correctly with error 1`] = `
<View>
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 12,
      }
    }
  >
    Test Label
  </Text>
  <Text
    testID="error-text"
  >
    Error message
  </Text>
  <TextInput
    multiline={false}
    numberOfLines={1}
    onChangeText={[MockFunction]}
    placeholder="Enter text"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "color": "#000",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
        "textAlignVertical": "top",
      }
    }
  />
</View>
`;
